### PR TITLE
[CFjs + Node] Rewrite/extend the CFjs.2 spec with Node Protocol

### PR DESCRIPTION
**Pull request by mvanderh**
- Add Node Protocol heading with methods and events
- Split AppDefinition into appId (on-chain address) and encodings
- Move install() to Provider
- Refer to Node Protocol methods and events for CFjs instance methods and events
- Leave anything with app manifest or proposing states TODO

### CODE REVIEW

File: packages/cf.js/API_REFERENCE.md [hunk 1]
```diff
@@ -1,82 +1,341 @@
...
+
+### `cfjs` Typescript Package
 
 - `Provider`
     - Properties
         - `nodeProvider: NodeProvider`
+            - Refer to `node-provider` package
     - Instance methods
         - `async getAppInstances(): AppInstance[]`
-        - `createAppFactory(appDefinition: AppDefinition): AppFactory`
```
Comment (ebryn) [inline packages/cf.js/API_REFERENCE.md]: why was `createAppFactory` removed?

Comment (mvanderh) [reply packages/cf.js/API_REFERENCE.md]: It wasn't necessary. Now you instantiate an AppFactory with a provider and an AppDefinition, which makes more sense.

File: packages/cf.js/API_REFERENCE.md [hunk 2]
```diff
@@ -1,82 +1,343 @@
...
         - `version`: semantic version of app definition contract
         - `definition: AppDefinition`
+
+Node Protocol
+=============
+
+Public Methods
+--------------
+
+#### getAppInstances
```
[the remark this answers is not in the document]
Comment (mvanderh) [reply packages/cf.js/API_REFERENCE.md]: Yeah good point, i thought the same when looking through the rendered version.

File: packages/cf.js/API_REFERENCE.md [hunk 3]
```diff
@@ -1,82 +1,343 @@
...
+            - [Node method](#rejectinstall)
     - Lifecycle
         - `on(eventType, callback: Function)`
-            - eventTypes
-                - `install(appInstance)`
-                - `rejectInstall(appInstance)`
-                - `updateState(appInstance, oldState, newState, action?)`
-                - `uninstall(appInstance)`
-                - `proposeState(appInstance, oldState, newState)`
+            - `install`
```
[the remark this answers is not in the document]
Comment (mvanderh) [reply packages/cf.js/API_REFERENCE.md]: Not sure i follow your question. The install event is fired on a successful app installation, meaning there exists a valid AppInstance for it.

File: packages/cf.js/API_REFERENCE.md [hunk 4]
```diff
@@ -1,82 +1,343 @@
...
+    - Extends [`AppInstance` data type](#appinstance)
     - Properties
-        - `id: AppInstanceID` — Identifier for this specific app instance
-        - `definition: AppDefinition`
-        - `terms: AppTerms`
         - `manifestUri: string`
+            - TODO
     - Instance methods
-        - `async applyAction(action: AppAction): AppState`
+        - `async takeAction(action: AppAction): Promise<AppState>`
```
[the remark this answers is not in the document]
Comment (mvanderh) [reply packages/cf.js/API_REFERENCE.md]: Well yeah, `appInstanceId` is implicit in the instance property `id` of the `AppInstance` class.

[the remark this answers is not in the document]
Comment (mvanderh) [reply packages/cf.js/API_REFERENCE.md]: Oh you're absolutely right, I forgot to add the action in the Node method.

[the remark this answers is not in the document]
Comment (cf19drofxots) [reply packages/cf.js/API_REFERENCE.md]: Yeah makes sense that it'd be implicit.

Just saying that the Node method should probably take the `action: AppAction` parameter as well (as indicated in the signature definition here, but not there)

File: packages/cf.js/API_REFERENCE.md [hunk 5]
```diff
@@ -1,82 +1,343 @@
...
-            - eventTypes
-                - `install(appInstance)`
-                - `rejectInstall(appInstance)`
-                - `updateState(appInstance, oldState, newState, action?)`
-                - `uninstall(appInstance)`
-                - `proposeState(appInstance, oldState, newState)`
+            - `install`
+                - [Node event](#install-event)
+                - Params: `(appInstance: AppInstance)`
+            - `rejectInstall`
```
[the remark this answers is not in the document]
Comment (cf19drofxots) [reply packages/cf.js/API_REFERENCE.md]: Stale

File: packages/cf.js/API_REFERENCE.md [hunk 6]
```diff
@@ -1,82 +1,344 @@
...
+
+### Method: `takeAction`
+
+Take action on current app state to advance it to a new state.
+
+Params:
+- `appInstanceId: string`
+    - Unique ID of the app instance 
+    - Action to take on the current state 
+- `action:`
```
Comment (cf19drofxots) [inline packages/cf.js/API_REFERENCE.md]: `action: AppAction`

File: packages/cf.js/API_REFERENCE.md [hunk 7]
```diff
@@ -1,82 +1,344 @@
...
+
+
+### Method: `takeAction`
+
+Take action on current app state to advance it to a new state.
+
+Params:
+- `appInstanceId: string`
+    - Unique ID of the app instance 
+    - Action to take on the current state
```
Comment (cf19drofxots) [inline packages/cf.js/API_REFERENCE.md]: Should be under the action section

**Pull request merged**